url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/PersistenceElement.java

```diff
@@ -0,0 +1,734 @@
+
+        if ( ((a >> 5) == 0x06) &&
+             ((b >> 6) == 0x02) ) {
+            c = new String(arr);
+        }
+        else {
+            System.out.println("VSDK: Unhandled UTF-8 binary encoding!");
+            System.out.println("  - Byte 0: " + a + " / " + (a >> 5) );
+            System.out.println("  - Byte 1: " + b + " / " + (b >> 6) );
+            return null;
+        }
+        return c;
+    }
+
+    public static String readUtf8Line(InputStream is) throws Exception
+    {
+        byte character[] = new byte[1];
+        char letter;
+        StringBuffer msg = new StringBuffer("");
+        byte a[] = new byte[2];
+
+        do {
+            if ( is.available() < 1 ) return "";
+            readBytes(is, character);
+            letter = (char)character[0];
+            if ( character[0] != '\n' && character[0] != '\r' &&
+                 ((letter >> 7) == 0) ) {
+                msg.append(letter);
+            }
+            else if ( character[0] != '\n' && character[0] != '\r' ) {
+                a[0] = character[0];
+                if ( is.available() >= 1 ) {
+                    readBytes(is, character);
+                    a[1] = character[0];
+                    msg.append(buildUtf8Char(a));
+                }
+            }
+        } while ( character[0] != '\n' );
+
+        return msg.toString();
+    }
+
+    public static String readAsciiLine(InputStream is) throws Exception
+    {
+        byte character[] = new byte[1];
+        char letter;
+        StringBuffer msg = new StringBuffer("");
+
+        do {
+            if ( is.available() < 1 ) return "";
+            readBytes(is, character);
+            letter = (char)character[0];
+            if ( character[0] != '\n' && character[0] != '\r' ) {
+                msg.append(letter);
+            }
+        } while ( character[0] != '\n' );
+
+        return msg.toString();
+    }
+
+    private static boolean isInSet(byte key, byte set[])
+    {
+        int i;
+
+        for ( i = 0; i < set.length; i++ ) {
+            if ( key == set[i] ) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static String readAsciiToken(InputStream is, byte separators[]) throws Exception
+    {
+        byte character[] = new byte[1];
+        char letter;
+        String msg = "";
+        int i;
+
+        do {
+            readBytes(is, character);
+            if ( !isInSet(character[0], separators) ) {
+                msg = msg + ((char)character[0]);
+            }
+        } while ( !isInSet(character[0], separators) );
+
+        return msg;
+    }
+
+    public static void
+    writeAsciiString(OutputStream writer, String cad)
+        throws Exception
+    {
+        byte arr[];
+        arr = cad.getBytes();
+        writer.write(arr, 0, arr.length);
+
+        byte end[] = new byte[1];
+        end[0] = '\0';
+        writer.write(end, 0, end.length);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 634:639, 646:651, 659:664, 672:677.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.